library/webpack-plugins/common-chunks-plugin.js

```diff
@@ -0,0 +1,376 @@
+    )
+  */
+
+  const chunkWithAllChunks = newChunks.find(({ chunks }) => chunks.length === affectedChunks.size)
+  if (chunkWithAllChunks) return addRuntime(chunkWithAllChunks)
+
+
+  const { exclusive, shared } = determineCoverage(newChunks, affectedChunks)
+  exclusive.forEach(addRuntime)
+
+  if (!shared.length) return
+
+  const runtimeChunkName = `common runtime (${shared.map(c => c.name).join(', ')})`
+  const runtimeChunk = addChunk(compilation, runtimeChunkName, { addRuntime: true })
+  createHierarchy({ parent: runtimeChunk, children: shared })
+}
+
+function addRuntime(chunk) {
+  chunk.entrypoints = [{ chunks: [chunk] }]
+}
+
+function determineAffectedChunks(newChunks) {
+  return newChunks.reduce(
+    (result, newChunk) => {
+      newChunk.chunks.forEach(chunk => {
+        const newChunks = result.get(chunk)
+        if (newChunks) newChunks.push(newChunk)
+        else result.set(chunk, [newChunk])
+      })
+      return result
+    },
+    new Map()
+  )
+}
+
+function determineCoverage(newChunks, affectedChunks) {
+  const shareCount = determineShareCounts(newChunks, affectedChunks)
+  /*
+    shareCount = Map(
+      [newChunk]: 0
+    )
+  */
+
+  return newChunks.reduce(
+    ({ exclusive, shared }, newChunk) => {
+      const isExclusive = shareCount.get(newChunk) === 1
+
+      if (isExclusive) exclusive.push(newChunk)
+      else shared.push(newChunk)
+
+      return { exclusive, shared }
+    },
+    { exclusive: [], shared: [] }
+  )
+}
+
+function determineShareCounts(newChunks, affectedChunks) {
+  return Array.from(affectedChunks.values()).reduce(
+    (result, newChunks) => {
+      newChunks.forEach(newChunk => {
+        const previousShareCount = result.get(newChunk) || 0
+        const newShareCount = newChunks.length
+
+        if (newShareCount > previousShareCount) result.set(newChunk, newShareCount)
+      })
+      return result
+    },
+    new Map()
+  )
+}
+
+/*
+  This function is not strictly necessary but allows us to, more easily catch edge cases
+  and changes to the webpack internals.
+
+  With this function we check if each original chunk that was affected by this module
+  has exactly 1 runtime in it's 'chunk-chain'
+*/
+function dubbelCheckRuntimes(newChunks) {
+  const affectedChunks = determineAffectedChunks(newChunks)
+  /*
+    affectedChunks = Map(
+      [chunk]: [newChunk1, ..., newChunkN]
+    )
+  */
+  Array.from(affectedChunks.entries()).forEach(
+    ([chunk, newChunks]) => {
+
+      if (chunk.hasRuntime()) error(
+        `Optimized chunk (${chunk.name}) has a runtime while it uses ${newChunks.length} shared chunks`
+      )
+
+      const newChunksWithRuntime = newChunks.filter(newChunk => newChunk.hasRuntime())
+      if (newChunksWithRuntime > 1) error(
+        `More than one shared chunk (used by a single chunk), ${newChunksWithRuntime.length} to be exact,\n` +
+        `has a runtime:\n` +
+        `${newChunksWithRuntime.map(c => c.name).join(', ')}`
+      )
```

Comment:
You can use line breaks in template literals
```js
error(`More than one shared chunk (used by a single chunk), ${newChunksWithRuntime.length} to be exact,
has a runtime:
${newChunksWithRuntime.map(c => c.name).join(', ')}`)

Reply:
Yeah, but that also includes the extra whitespace. I should probably use that `stripMargin` regex. I might even define that on the String prototype so we can easily use it in the build tool.